File: 0400-nth-digit/0400-nth-digit.py
class Solution:
    def findNthDigit(self, n):
        digit_len = 1
        count = 9
        start = 1

        while n > digit_len * count:
            n -= digit_len * count
            digit_len += 1
            count *= 10
            start *= 10

        num = start + (n - 1) // digit_len
        digit_index = (n - 1) % digit_len

        return int(str(num)[digit_index])

File: 0400-nth-digit/0400-nth-digit.cpp
class Solution {
public:
    int findNthDigit(int n) {
        long long digits = 1;
        long long count = 9;
        long long start = 1;

        // Find the range containing the nth digit
        while (n > digits * count) {
            n -= digits * count;
            digits++;
            count *= 10;
            start *= 10;
        }

        // Find the actual number
        long long num = start + (n - 1) / digits;

        // Find the digit inside that number
        int index = (n - 1) % digits;

        string s = to_string(num);

        return s[index] - '0';
    }
};